src/apply.jl

```diff
@@ -25,16 +25,49 @@ function ITensors.apply(
     if ortho
       ψ = orthogonalize(ψ, v⃗[1])
     end
-    oψᵥ = apply(o, ψ[v⃗[1]] * ψ[v⃗[2]])
-    ψᵥ₁, ψᵥ₂ = factorize(
-      oψᵥ, inds(ψ[v⃗[1]]); cutoff, maxdim, tags=ITensorNetworks.edge_tag(e)
+
+    #Check whether to do a memory efficient QR first (do both sites combined for now, although we could split if we wish)
+    #if no maxdim is provided just do it based on inds counting, if it is provided do it based on whichever takes less memory
+
+    v1_siteinds = setdiff(
+      inds(ψ[v⃗[1]]), flatten(unique(inds.([ψ[v] for v in neighbors(ψ, v⃗[1])])))
     )
+    v2_siteinds = setdiff(
+      inds(ψ[v⃗[2]]), flatten(unique(inds.([ψ[v] for v in neighbors(ψ, v⃗[2])])))
+    )
+    d1, d2 = prod(dim.(v1_siteinds)), prod(dim.(v2_siteinds))
+    χv1, χv2 = prod(
+      dim.(uniqueinds(uniqueinds(ψ[v⃗[1]], ψ[v⃗[2]]), inds(ψ[v⃗[1]]; tags="Site")))
+    ),
+    prod(dim.(uniqueinds(uniqueinds(ψ[v⃗[2]], ψ[v⃗[1]]), inds(ψ[v⃗[2]]; tags="Site"))))
+    if (maxdim == nothing && length(inds(ψ[v⃗[1]])) + length(inds(ψ[v⃗[2]])) <= 6) ||
+      (maxdim != nothing && maxdim * maxdim * d1 * d2 >= χv1 * χv2)
+      oψᵥ = apply(o, ψ[v⃗[1]] * ψ[v⃗[2]])
+      ψᵥ₁, ψᵥ₂ = factorize(
+        oψᵥ, inds(ψ[v⃗[1]]); cutoff, maxdim, tags=ITensorNetworks.edge_tag(e)
+      )
+    else
+      Qv1, Rv1 = factorize(
+        ψ[v⃗[1]], uniqueinds(uniqueinds(ψ[v⃗[1]], ψ[v⃗[2]]), inds(ψ[v⃗[1]]; tags="Site"))
+      )
+      Qv2, Rv2 = factorize(
+        ψ[v⃗[2]], uniqueinds(uniqueinds(ψ[v⃗[2]], ψ[v⃗[1]]), inds(ψ[v⃗[2]]; tags="Site"))
+      )
+      Rv1_new, Rv2_new = factorize(
+        noprime(Rv1 * o * Rv2), inds(Rv1); cutoff, maxdim, tags=ITensorNetworks.edge_tag(e)
+      )
+      ψᵥ₁ = Qv1 * Rv1_new
+      ψᵥ₂ = Qv2 * Rv2_new
+    end
+
```

Comment:
To clarify the code structure, this is all still a "simple update" (i.e. no environment) but uses QR factorizations to improve performance?

It seems better to split this part into a separate function which you call inside this function. Also, shouldn't this same code and logic apply to the "full update"?

It seems like this is all pointing towards just having a single `apply` code that skips some steps in the "simple update" case, for example has a special case where the environment is empty so doesn't do an optimization and just a factorization.